did/key/key.go

```diff
@@ -90,176 +124,265 @@ func CreateDIDKey(kt crypto.KeyType, publicKey []byte) (*DIDKey, error) {
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "could not convert ed25519 public key to x25519")
+		}
+		id, x25519Key, err := x25519KeyAndID(vm.Controller, ed25519PubKey.(ed25519.PublicKey))
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "could not generate x25519 key and id")
+		}
+		verificationMethod, vmErr = did.ConstructJWKVerificationMethod(id, vm.Controller, x25519Key, crypto.X25519)
+		verificationMethodSet = []did.VerificationMethodSet{id}
+	} else {
+		verificationMethodSet = []did.VerificationMethodSet{vm.ID}
 	}
+	return verificationMethod, verificationMethodSet, vmErr
+}
 
-	cryptoKeyType, err := did.MultiCodecToKeyType(multiCodecValue)
+func x25519KeyAndID(id string, ed25519PubKey ed25519.PublicKey) (string, []byte, error) {
+	if len(ed25519PubKey) != ed25519.PublicKeySize {
+		return "", nil, errors.New("ed25519 public key is not the right size")
+	}
+	x25519Key := ed2curve25519.Ed25519PublicKeyToCurve25519(ed25519PubKey)
+	keyAgreementDIDKey, err := CreateDIDKey(crypto.X25519, x25519Key)
 	if err != nil {
-		return nil, "", "", errors.Wrap(err, "codec to key type")
+		return "", nil, errors.Wrap(err, "could not multibase encode x25519 public key")
 	}
-
-	return pubKeyBytes, ldKeyType, cryptoKeyType, nil
-}
-
-// decode public key with type
-func decodePublicKeyWithType(data []byte) ([]byte, cryptosuite.LDKeyType, error) {
-	encoding, decoded, err := multibase.Decode(string(data))
+	suffix, err := keyAgreementDIDKey.Suffix()
 	if err != nil {
-		return nil, "", errors.Wrap(err, "failed to decode public key")
+		return "", nil, errors.Wrap(err, "could not get suffix from did:key")
 	}
+	return id + "#" + suffix, x25519Key, nil
+}
 
-	if encoding != did.Base58BTCMultiBase {
-		err := fmt.Errorf("expected %d encoding but found %d", did.Base58BTCMultiBase, encoding)
-		return nil, "", err
+func processExpansionOptions(opts ...Option) (cryptosuite.LDKeyType, bool, error) {
+	publicKeyFormat := cryptosuite.JSONWebKey2020Type
+	enableEncryptionKeyDerivation := true
+	var ok bool
+	for _, opt := range opts {
+		switch opt.Name {
+		case "":
+			continue
+		case PublicKeyFormatOption:
+			publicKeyFormat, ok = opt.Value.(cryptosuite.LDKeyType)
+			if !ok {
+				return "", false, fmt.Errorf("invalid public key format option type: %T", opt.Value)
+			}
+
+		case EnableEncryptionKeyDerivationOption:
+			enableEncryptionKeyDerivation, ok = opt.Value.(bool)
+			if !ok {
+				return "", false, fmt.Errorf("invalid enable encryption key derivation option type: %T", opt.Value)
+			}
+		default:
+			return "", false, fmt.Errorf("invalid option: %s", opt.Name)
+		}
 	}
+	return publicKeyFormat, enableEncryptionKeyDerivation, nil
+}
 
-	// n = # bytes for the int, which we expect to be two from our multicodec
-	multiCodec, n, err := varint.FromUvarint(decoded)
-	if err != nil {
-		return nil, "", errors.Wrap(err, "failed to decode public key from varint")
+func IsSupportedDIDKeyType(kt crypto.KeyType) bool {
```

Comment:
Should this use a map instead of doing a linear pass?

Reply:
I wanted to avoid having two representations of the official list